addons/mass_mailing/models/mail_mail.py

```diff
@@ -60,28 +59,29 @@ def _send_prepare_body(self):
                 '<img src="%s"/>' % tracking_url,
                 plaintext=False,
             )
-
-        body = self.env['mail.render.mixin']._replace_local_links(body)
-
         return body
 
-    def _send_prepare_values(self, partner=None):
-        # TDE: temporary addition (mail was parameter) due to semi-new-API
-        res = super(MailMail, self)._send_prepare_values(partner)
-        if self.mailing_id and res.get('body') and res.get('email_to'):
-            base_url = self.mailing_id.get_base_url()
-            emails = tools.email_split(res.get('email_to')[0])
-            email_to = emails and emails[0] or False
+    def _prepare_outgoing_list(self):
+        """ Update mailing specific links to add tracking based on res_id """
+        email_list = super()._prepare_outgoing_list()
+        if not self.res_id or not self.mailing_id:
+            return email_list
+        base_url = self.mailing_id.get_base_url()
+        for email_values in email_list:
+            if tools.is_html_empty(email_values['body']) or not email_values['email_to']:
+                continue
+            emails = tools.email_split(email_values['email_to'][0])
+            email_to = emails[0] if emails else False
 
             urls_to_replace = [
                 (base_url + '/unsubscribe_from_list', self.mailing_id._get_unsubscribe_url(email_to, self.res_id)),
                 (base_url + '/view', self.mailing_id._get_view_url(email_to, self.res_id))
```

Comment:
> we don't have any insight it is important to modify it.

We are calling `_get_unsubscribe_url` **and** `_get_view_url` for each `email_values`, even when it is not necessary.

But I guess it can come in another PR :)